instrumentation/bunny/lib/opentelemetry/instrumentation/bunny/patch_helpers.rb

```diff
@@ -0,0 +1,63 @@
+# frozen_string_literal: true
+
+# Copyright The OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+module OpenTelemetry
+  module Instrumentation
+    module Bunny
+      # The PatchHelper module provides functionality shared between patches.
+      #
+      # For additional details around trace messaging semantics
+      # See https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/messaging.md#messaging-attributes
+      module PatchHelpers
+        def self.with_send_span(tracer, exchange, routing_key, &block)
+          attributes = basic_attributes(exchange, routing_key)
+          destination = destination_name(exchange, routing_key)
+
+          tracer.in_span("#{destination} send", attributes: attributes, kind: :producer, &block)
+        end
+
+        def self.with_process_span(tracer, delivery_info, properties, &block)
+          with_consumer_span(tracer, 'process', delivery_info, properties, &block)
+        end
+
+        def self.with_receive_span(tracer, delivery_info, properties, &block)
+          with_consumer_span(tracer, 'receive', delivery_info, properties, &block)
+        end
+
+        def self.with_consumer_span(tracer, operation, delivery_info, properties, &block)
+          exchange = delivery_info[:exchange]
+          routing_key = delivery_info[:routing_key]
+          attributes = basic_attributes(exchange, routing_key)
+          destination = destination_name(exchange, routing_key)
+
+          parent_context = OpenTelemetry.propagation.text.extract(properties[:headers])
+          span_context = OpenTelemetry::Trace.current_span(parent_context).context
+          links = [OpenTelemetry::Trace::Link.new(span_context)] if span_context.valid?
+
+          OpenTelemetry::Context.with_current(parent_context) do
+            tracer.in_span("#{destination} #{operation}", links: links, attributes: attributes, kind: :consumer, &block)
+          end
+        end
+
+        def self.destination_name(exchange, routing_key)
+          [exchange, routing_key].compact.join('.')
+        end
+
+        def self.basic_attributes(exchange, routing_key)
+          attributes = {
+            'messaging.system' => 'rabbitmq',
+            'messaging.destination' => exchange,
+            'messaging.destination_kind' => 'topic',
```

Comment:
I think this should typically be `queue` for RabbitMQ. If a topic exchange is used, then this should be `topic`, but that doesn't look like the default use-case.

Reply:
I'm not entirely sure as even the default exchange model used in the example is topic based under the hood https://www.rabbitmq.com/tutorials/amqp-concepts.html#exchange-default (we also only use topic exchanges at my company)

I guess this could be set based on the exchange type https://github.com/ruby-amqp/bunny/blob/master/lib/bunny/exchange.rb#L22 `queue` for direct, `topic` otherwise?

Reply:
> I guess this could be set based on the exchange type https://github.com/ruby-amqp/bunny/blob/master/lib/bunny/exchange.rb#L22 `queue` for direct, `topic` otherwise?

👍

Reply:
Fixed.